Automattic/wp-calypso
# Tools: New package 'effective-module-tree'

#### Changes proposed in this Pull Request

* Adds a new package `effective-module-tree`. This package is a CLI tool meant to be used to assists moving to a different package manger, by generating a logical dependency tree (with all hoisted/deduped packages "resolved") so we can compare trees generated by different package managers.

#### Testing instructions

* Clone the repo and install all dependencies (`npm run distclean; PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=1 npm ci`)
* Run `./node_modules/.bin/effective-module-tree` and check that the output makes sense

It works on any project as long as you run it in a directory with `package.json` and `node_modules`. Can be used to analyse other projects outside wp-calypso.

#### Example

This is the output of using the tool to analyse itself:

```
└─ @automattic/effective-module-tree@0.0.1
   ├─ debug@4.1.1
   │  └─ ms@2.1.2
   ├─ globby@11.0.0
   │  ├─ array-union@2.1.0
   │  ├─ dir-glob@3.0.1
   │  │  └─ path-type@4.0.0
   │  ├─ fast-glob@3.2.2
   │  │  ├─ @nodelib/fs.stat@2.0.3
   │  │  ├─ @nodelib/fs.walk@1.2.4
   │  │  │  ├─ @nodelib/fs.scandir@2.1.3
   │  │  │  │  ├─ @nodelib/fs.stat@2.0.3
   │  │  │  │  └─ run-parallel@1.1.9
   │  │  │  └─ fastq@1.6.1
   │  │  │     └─ reusify@1.0.4
   │  │  ├─ glob-parent@5.1.0
   │  │  │  └─ is-glob@4.0.1
   │  │  │     └─ is-extglob@2.1.1
   │  │  ├─ merge2@1.3.0
   │  │  ├─ micromatch@4.0.2
   │  │  │  ├─ braces@3.0.2
   │  │  │  │  └─ fill-range@7.0.1
   │  │  │  │     └─ to-regex-range@5.0.1
   │  │  │  │        └─ is-number@7.0.0
   │  │  │  └─ picomatch@2.2.1
   │  │  └─ picomatch@2.2.1
   │  ├─ ignore@5.1.4
   │  ├─ merge2@1.3.0
   │  └─ slash@3.0.0
   └─ object-treeify@1.1.23
```
